ssl-hep/ServiceX_App
# Report Fatal Errors from DID finder and other Services

# Problem
Fatal errors can be generated in downstream ServiceX services. These get logged, but are never reported to the user.

Partial solution to [ServiceX issue 153](https://github.com/ssl-hep/ServiceX/issues/153)

# Approach
1. Change the body of the transform status POST endpoint to include:
a. `Severity` can be _debug, info, warn,_ or _fatal_
b. `info` details on the error or status message
c. `source` which service the status report came from
2. Update the `TransformRequest` table to hold the status and failure description
3. Update the GET transform request endpoint to return the status and the failure reason
4. Update the transform status request endpoint to return the current status so the client can detect a failure as soon as it is attached to the transform record.
5. This required changes to the database. Up to now, whenever we deploy the application, the database is completely dropped and erased. Now that we are using persistent volumes on River, we need to consider database migration. 
a. We introduced alelmbic database migration tooling to the flask server. 
b. Database migrations are performed in `boot.sh` with the command: `FLASK_APP=servicex/app.py flask db upgrade`
c. Had to add an `app.py` for the flask command line tooling to work


# How to test
1. Submit a transform request with an non-existent DID
2. Look at the results from the `{{host}}/servicex/transformation/:request_id?` endpoint and see that the `status` is set to _Submitted_ or _Running_
3. Do the same for the `{{host}}/servicex/transformation/:request_id/status` endpoint
4. Eventually in the app log you will see the fatal error reported:
```
10.1.7.76 - - [27/Jul/2020:15:06:54 +0000] "POST /servicex/internal/transformation/8cd40723-2cb9-4483-b71c-219e870ee2ba/status HTTP/1.1" 200 5 "-" "python-requests/2.21.0"
+--------------------------------------------+

  ______   _______       _        ______ _____  _____   ____  _____
 |  ____/\|__   __|/\   | |      |  ____|  __ \|  __ \ / __ \|  __ \
 | |__ /  \  | |  /  \  | |      | |__  | |__) | |__) | |  | | |__) |
 |  __/ /\ \ | | / /\ \ | |      |  __| |  _  /|  _  /| |  | |  _  /
 | | / ____ \| |/ ____ \| |____  | |____| | \ \| | \ \| |__| | | \ \
 |_|/_/    \_\_/_/    \_\______| |______|_|  \_\_|  \_\\____/|_|  \_\
            
+ Fatal error reported for 8cd40723-2cb9-4483-b71c-219e870ee2ba from DID Finder
DID Not found mc15_13TeV:mc15_13TeV.361106.PowhegPythia8EvtGen_AZNLOCTEQ6L1_Zee.merge.DAOD_STDM3.e3601_s2576_s2132_r6630_r6264_p2363_tid05630052_00XXXX
+--------------------------------------------+
```
5. Check that the transform request now shows a `Fatal` status with DID not found info failure in the `failure-info` property.
6. Check that the transform status endpoint shows the `Fatal` status.
